python/ray/dag/compiled_dag_node.py

```diff
@@ -1213,34 +1215,161 @@ def _preprocess(self) -> None:
+        """
+
+        # First, initialize communicators that are passed in by the user.
+        for communicator, type_hints in self._communicator_to_type_hints.items():
+            communicator_id = _init_communicator(
+                communicator.get_actor_handles(),
+                communicator,
+                self._overlap_gpu_communication,
+            )
+            for type_hint in type_hints:
+                type_hint.set_communicator_id(communicator_id)
+
+        # Then, create communicators for collective operations.
+        # Reuse an already created communicator for the same set of actors.
+        for collective_op in self._pending_collective_ops:
+            if not self._create_default_communicator:
+                raise ValueError(
+                    "Communicator creation is not allowed for collective operations."
+                )
+            actors = frozenset(collective_op.actor_handles)
+            communicator_id = collective_op.init_communicator(
+                self._actors_to_created_communicator_id.get(actors, None)
+            )
+            if actors not in self._actors_to_created_communicator_id:
+                self._actors_to_created_communicator_id[actors] = communicator_id
+
+        # Finally, create a communicator for P2P operations.
+        # Reuse an already created collective op communicator when p2p actors
+        # are a subset of the actors in the collective op communicator.
+        p2p_communicator_id = None
+        if self._pending_p2p_communicator_actors:
+            for (
+                actors,
+                communicator_id,
+            ) in self._actors_to_created_communicator_id.items():
+                if self._pending_p2p_communicator_actors.issubset(actors):
+                    p2p_communicator_id = communicator_id
+                    break
+            if p2p_communicator_id is None:
+                p2p_communicator_id = _init_communicator(
+                    list(self._pending_p2p_communicator_actors),
+                    None,
+                    self._overlap_gpu_communication,
+                )
+            for dag_node in self._pending_p2p_communicator_dag_nodes:
+                dag_node.type_hint.set_communicator_id(p2p_communicator_id)
+
+    def _track_communicator_usage(
+        self,
+        dag_node: "ray.dag.DAGNode",
+        actors: Set["ray.actor.ActorHandle"],
+        collective_op: bool = False,
+    ) -> None:
+        """
+        Track the usage of a communicator.
+
+        This method first determines the communicator to use: if a custom
+        communicator is specified, use it; if not and a default communicator
+        is available, use it; otherwise, it records necessary information to
+        create a new communicator later.
+
+        This method also performs validation checks on the passed-in communicator.
+
+        Args:
+            dag_node: The DAG node that uses the communicator.
```

Comment:
It’s not very clear to me what “The DAG node that uses the communicator” means when I first read it. Does it include both the writer DAG node and the reader DAG nodes?

Maybe we can add more details such as the DAG node with `with_tensor_transport`?